node/bindnode/node.go

```diff
@@ -0,0 +1,1103 @@
+			finish:     w.finish,
+		}, nil
+	}
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "BeginMap",
+		// TODO
+	}
+}
+
+func (w *_assembler) BeginList(sizeHint int64) (ipld.ListAssembler, error) {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeList:
+		val := w.nonPtrVal()
+		return &_listAssembler{
+			schemaType: typ,
+			val:        val,
+			finish:     w.finish,
+		}, nil
+	}
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "BeginList",
+		// TODO
+	}
+}
+
+func (w *_assembler) AssignNull() error {
+	if !w.nullable {
+		return ipld.ErrWrongKind{
+			TypeName:   w.schemaType.Name().String(),
+			MethodName: "AssignNull",
+			// TODO
+		}
+	}
+	w.val.Set(reflect.Zero(w.val.Type()))
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_assembler) AssignBool(bool) error {
+	panic("TODO: AssignBool")
+}
+
+func (w *_assembler) AssignInt(i int64) error {
+	if w.kind() != ipld.Kind_Int {
+		return ipld.ErrWrongKind{
+			TypeName:        w.schemaType.Name().String(),
+			MethodName:      "AssignInt",
+			AppropriateKind: ipld.KindSet{ipld.Kind_Int},
+			ActualKind:      w.kind(),
+		}
+	}
+	w.nonPtrVal().SetInt(i)
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_assembler) AssignFloat(float64) error {
+	panic("TODO: AssignFloat")
+}
+
+func (w *_assembler) AssignString(s string) error {
+	if w.kind() != ipld.Kind_String {
+		return ipld.ErrWrongKind{
+			TypeName:        w.schemaType.Name().String(),
+			MethodName:      "AssignString",
+			AppropriateKind: ipld.KindSet{ipld.Kind_String},
+			ActualKind:      w.kind(),
+		}
+	}
+	w.nonPtrVal().SetString(s)
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_assembler) AssignBytes(p []byte) error {
+	if w.kind() != ipld.Kind_Bytes {
+		return ipld.ErrWrongKind{
+			TypeName:        w.schemaType.Name().String(),
+			MethodName:      "AssignBytes",
+			AppropriateKind: ipld.KindSet{ipld.Kind_Bytes},
+			ActualKind:      w.kind(),
+		}
+	}
+	w.nonPtrVal().SetBytes(p)
+	return nil
+}
+
+func (w *_assembler) AssignLink(link ipld.Link) error {
+	newVal := reflect.ValueOf(link)
+	if !newVal.Type().AssignableTo(w.val.Type()) {
+		return ipld.ErrWrongKind{
+			TypeName:   w.schemaType.Name().String(),
+			MethodName: "AssignLink",
+			// TODO
+		}
+	}
+	w.nonPtrVal().Set(newVal)
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_assembler) AssignNode(node ipld.Node) error {
+	// TODO: does this ever trigger?
+	// newVal := reflect.ValueOf(node)
+	// if newVal.Type().AssignableTo(w.val.Type()) {
+	// 	w.val.Set(newVal)
+	// 	return nil
+	// }
+	switch node.Kind() {
+	case ipld.Kind_Map:
+		itr := node.MapIterator()
+		am, err := w.BeginMap(-1) // TODO: length?
```

Comment:
I think I would use node.Length here, yeah.

Also: I've been considering extracting this whole hunk below to a generic function like, say, `ipld.BlindCopyMap`, because it does come up in a couple places now.  (The other big example is similarly, in the AssignNode methods emitted by codegen, this logic exists as a fallback path right after trying to do direct copy for known types.)  So if that looks reasonable to extract, go for it.

Reply:
I sorta left these TODOs as reminders for future optimizations, and as said before, I have no benchmarks. From what you say, we might dedup this anyway :)

I'll add this refactor as another TODO.